improve queues, (one effect: improve code completion)

an-dr-eas-k opened the pull request:
After updating to the recent 1.23.x versions I had severe performance issues on code completion. 

I investigated that and made the following changes: 
* Completion, CompletionResolve, and AutoComplete should be equally prioritized (add to normal queue)
* remove request from _waiting queue when cancelRequest is called and call onError on the request to reject() any awaited promises.
* add logging (debugMode=true in main.ts) for canceled requests and add logging details to RequestQueue internals.

The second bullet point addresses a deferred queue that stays always full when requests are cancelled. A symptom for that are many log entries of the kind: 
``` 
Problem invoking 'GetCodeActions' on OmniSharp server: Error: Pending request cancelled: /v2/getcodeactions
```

filipw commented on src/omnisharp/prioritization.ts:
```diff
@@ -13,6 +13,8 @@ const priorityCommands = [
 const normalCommands = [
+    protocol.Requests.Completion,
+    protocol.Requests.CompletionResolve,
     protocol.Requests.AutoComplete,
```
this makes sense to me, thanks. Also autocomplete is no longer used, so I guess we could clean up all the references to it @JoeRobich @333fred

333fred replied:
Looks like this is the only remaining reference, it and the constant can be deleted.

333fred replied:
@an-dr-eas-k would you mind remove this reference?

an-dr-eas-k commented:
For me, especially the other change, calling dequeue when a request is canceled is most important. Please note the log entries that I entered to debug the behaviour in 954ce95. I found out, the deferred queue is considered full just when two requests are canceled but remain in the _waiting queue forever. Then the queue does no longer accept further requests. This especially had an impact on code completion in the last versions. But also, Unit-test classes loose the "RunTest" links, Code Structure gets lost in the outline panel and so on. 

The code I am working on has lots of automapper statements which made trouble when calling codeactions couple of times that were subsequently canceled, some already _waiting. And suddenly all requests that are prioritized to deferred queue didn't reach out for the omnisharp server.

I am not sure, if adding the id to the requests is proper, but at least solves a quick handling in the dequeue method.

I think this fix addresses at least some of:
#4292 
#4289
#4311 
#4080
#4047
#4072
#3421
#4065

333fred commented on src/omnisharp/server.ts:
```diff
@@ -705,7 +705,8 @@ export class OmniSharpServer {
         const id = OmniSharpServer._nextId++;
-
+        request.id = id;
+        
```
Nit: some extra whitespace.

333fred reviewed (dismissed):
Overall looks good. We can delete the autocomplete entry and the definition in protocol.ts as well.

an-dr-eas-k commented:
Sure. I needed to do these adjustments because I realized the OmnisharpWorkspaceSymbolProvider (ctrl-t) also needs request.onError to cancel the awaited task when it is in the waiting queue. 

Essentially what changed in the last 4 commits are two things:
1. I added the OmnisharpServerRequestCancelled BaseEvent and did some small changes to other BaseEvents to get more logging for canceled requests in the output panel when debug=true in main.ts.
2. I moved the request.onError call one level higher to be called independent of the queueStatus the request is in. Originally, onError was only called for requests in the _pending queue. 

Cheers.

333fred commented on src/omnisharp/EventType.ts:
```diff
@@ -82,7 +82,8 @@ export enum EventType {
     DotNetTestDebugInContextStart = 77,
-    TelemetryErrorEvent = 78
+    TelemetryErrorEvent = 78,
+    OmnisharpServerRequestCancelled = 79
```
Nit: consider leaving a trailing comma here, so that future PRs that add elements don't have to modify this line.

333fred commented on src/omnisharp/loggingEvents.ts:
```diff
@@ -115,17 +115,22 @@ export class OmnisharpServerUnresolvedDependencies implements BaseEvent {
 export class OmnisharpServerProcessRequestStart implements BaseEvent {
     type = EventType.OmnisharpServerProcessRequestStart;
-    constructor(public name: string) { }
+    constructor(public name: string, public slots: number) { }
```
`slots` isn't a particularly descriptive name. Can you please rename it to something more descriptive?

333fred reviewed (changes requested):
Thanks. This is mostly looking good, just a couple of small pieces of feedback.

333fred reviewed (approved):
LGTM. @JoeRobich for a second review.

JoeRobich reviewed (approved):
Thanks!
